Fix for Configuration Binder Source Gen duplicate field names

alex-lutsenko opened the pull request:
Configuration Binder Source Generator generates duplicate field names
where the same class name is present in different namespaces.

Fix #119458

Commit:
Fix for Configuration Binder Source Gen duplicate field names

Configuration Binder source generator generates duplicate field names
where the same class name is present in different namespaces.

Fix #119458

alex-lutsenko commented:
@tarekgh - please review this.

alex-lutsenko commented:
I have run `dotnet build /p:UpdateBaselines=true` for InterceptableAttributeVersion = 0 and it has resulted in every Version 0 file being updated (`src\libraries\Microsoft.Extensions.Configuration.Binder\tests\SourceGenerationTests\Baselines` ). The changes are not related to my fix. Looks like these baselines are out of date.

Please advise if I should include updated Version 0 baselines in my PR.

tarekgh commented on src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Parser/Extensions.cs:
```diff
@@ -89,35 +89,56 @@ public static void RegisterCacheEntry<TKey, TValue, TEntry>(this Dictionary<TKey
+            ToIdentifierCompatibleSubstringBuilder(type, sb);
+
+            sb.Replace(".", "").Replace("[", "").Replace("]", "");
+
```
Wouldn't that can still generate a conflict names? A.BC and AB.C for instance?

alex-lutsenko replied:
Yes, for objects with no namespace. Preparing another build.

tarekgh replied:
Would make sense to append a number to the created label? like use https://github.com/dotnet/runtime/blob/e4e0810cf1812d7ed46c457f37c864404170ec8a/src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Emitter/Helpers.cs#L200? At that time, and no need to include the namespace either.

alex-lutsenko replied:
I have checked in my latest attempt at `ToIdentifierCompatibleSubstring`, which includes the fix for "A.BC vs AB.C".

Commit:
Replace "." with "__" to solve Namespace.A vs
NamespaceA name clashes. "SymbolDisplayMiscellaneousOptions.UseSpecialTypes" - is no longer a problem.
Added "___" as generic type delimiter to solve A.BC vs AB.c name clashes.
Shortened maximum identifier length to slightly improve code readability.

Commit:
Regenerate InterceptableAttributeVersion=0 baselines

alex-lutsenko committed:
Merge branch 'main' into issue-119458

tarekgh reviewed (approved):
Thanks @alex-lutsenko!

tarekgh commented:
>  I do not see my fixes in the latest (10.0.0-rtm.25479.115) for Microsoft.Extensions.Configuration.Binder. Am I missing something? Is it not going to be released in .NET 10?

No, this is scheduled for .NET 11. Including it in .NET 10 would be risky since the fix hasn’t been released in any preview, and we can’t guarantee that it won’t introduce side effects.
